text/0095-cognito-construct-library/working-backwards-readme.md

```diff
@@ -0,0 +1,600 @@
+  deliveryMediums: [ ContactMethods.EMAIL, ContactMethods.PHONE ],
+  attributes: {
+    [StandardAttrs.EMAIL]: 'support@awesomeapp.com',
+    callingcode: 12,
+  },
+  forceAliasCreation: true,
+  resendInvitation: true,
+});
+```
+
+The property `userName` is the name of the user and is mandatory.
+
+The property `deliveryMedium` is how the user will be contacted to sign up, upon first creation. Both email and sms are
+available here.
+
+The property `attributes` specifies the list of standard and custom [attributes](#attributes) as part of the user.
+Values of the `StandardAttrs` enum can be used against `name` to reference standard attributes. If one or both contact
+methods are set with `deliveryMedium`, this attribute is required to be set as part of `attributes`.
+
+`forceAliasCreation` can be used only when the user pool is configured with either email or phone as a verified contact
+method. When this is set, any existing user with the same email address or phone number will have its alias migrated to
+this user.
+
+When the `resendInvitation` property is set to true, Cognito will resend the invitation message for a user that already
+exists but has not signed up.
+
+Users can be created for imported user pools by using the `UserPoolUser` construct, as so -
+
+```ts
+const userpool = UserPool.fromUserPoolAttributes(this, 'myuserpool', { ... });
+
+new UserPoolUser(this, userpool, {
+  userPool: userpool,
+  // ...
+  // all of the same properties as above
+  // ...
+});
+```
+
+> These are the defaults that will be part of tsdoc, and not part of the README -
+> * deliveryMedium: ContactsMethod.PHONE
+> * attributes: none
+> * forceAliasCreation: false
+> * resendInvitation: false
+
+### Groups
+
+User pools can be configured with groups into which users are added. Groups is the primary mechanism in Cognito for
+permission assignment.
+
+Add a new group to the user pool using the `createGroup()` API, as so -
+
+```ts
+const userpool = new UserPool(this, 'myuserpool', { ... });
+
+const customerrole = new iam.Role(this, 'customer-role', { ... });
+
+userpool.createGroup(this, 'myuserpool-customers', {
+  groupName: 'customers',
+  description: 'group for myawesomeapp customers',
+  role: customerrole
+});
+```
+
+The required `groupName` and the optional `description` properties are the identifiers for this group.
+
+The `role` property refers to the IAM role whose permissions that the users in this group are able to assume.
+
+When several different groups are defined for the same user pool, they can be ordered to have a precendence. Precedence
+will determine which IAM role's permission will be available for a user when they are part of two different groups. See
+[CreateGroup
+API](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateGroup.html#CognitoUserPools-CreateGroup-request-Precedence)
+for more details.
+
+Use the `takesPrecendenceOver()` API to set the precedence of one group over the other.
```

Comment:
would it be possible to set the precedence on the user pool itself?
this seems like a really easy way to create a circular graph. Or rather will we detect them to fail faster?

```ts
 groupA.takesPrecedenceOver(groupB);
 ...
 groupB.takesPrecedenceOver(groupA);
```

Reply:
I think we should detect it and fail faster. I'd like to play around with this after implementation to see how the final outcomes looks and adjust accordingly. 

This is just at a proposal level here, as in, let me know if this sounds like a good, bad or if there's a better idea. Smaller adjustments can be reviewed in the actual implementation PR.

Reply:
agreed, just capturing thoughts that come to mind.  failing faster sounds like the right customer experience to me